Enable text indicator styling for specified ranges.

megangardner opened the pull request:
#### a35a92c62f0e85ebd505ed1fac53febe94facb6f
<pre>
Enable text indicator styling for specified ranges.
<a href="https://bugs.webkit.org/show_bug.cgi?id=272293">https://bugs.webkit.org/show_bug.cgi?id=272293</a>
<a href="https://rdar.apple.com/125359006">rdar://125359006</a>

Reviewed by Wenson Hsieh.

Enable text indicator styling for specified ranges.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _enableTextIndicatorStylingAfterElementWithID:]):
(-[WKWebView _disableTextIndicatorStylingWithUUID:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::enableTextIndicatorStyleAfterElementWithID):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::getTextIndicatorForID): Deleted.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getTextIndicatorForID):
(WebKit::WebPage::enableTextIndicatorStyleAfterElementWithID):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/277284@main">https://commits.webkit.org/277284@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff59882774bad13e73bab8a8a1fc0c22217f4ce3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47202 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26381 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49849 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49885 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43251 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31722 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23841 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38447 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47783 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23882 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40685 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19760 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21277 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5246 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43579 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42248 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51761 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22231 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18602 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45743 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23504 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44752 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10411 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24291 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23222 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -322,6 +348,26 @@
+    // FIXME: move the start of the range to be after the element with the ID listed.
+
+    RefPtr frame = m_page->checkedFocusController()->focusedOrMainFrame();
+    if (!frame)
```
all of these early returns should also have `ASSERT_NOT_REACHED` before them

rr-codes commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -322,6 +348,26 @@
+
+    RefPtr frame = m_page->checkedFocusController()->focusedOrMainFrame();
+    if (!frame)
+        return;
```
some extra newlines after the early returns would also help for readability

rr-codes commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
```diff
@@ -73,7 +73,7 @@ class UnifiedTextReplacementController final {
     void updateStateForSelectedReplacementIfNeeded();
 
-    void getTextIndicatorForID(const WTF::UUID&, CompletionHandler<void(std::optional<WebCore::TextIndicatorData>&&)>&&);
+    RefPtr<WebCore::Range> rangeForUUID(const WTF::UUID uuid) { return m_contextRanges.get(uuid); }
```
this should be `const WTF::UUID& uuid`

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -379,6 +379,7 @@ struct PerWebProcessState {
 
 - (void)_textReplacementSession:(NSUUID *)sessionUUID updateState:(WebKit::WebTextReplacementDataState)state forReplacementWithUUID:(NSUUID *)replacementUUID;
 - (void)_removeTextIndicatorStyleForID:(NSUUID *)uuid;
+
```
Nit - stray newline (either remove this, or add an additional newline right below the `ENABLE(UNIFIED_TEXT_REPLACEMENT)`)

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -322,6 +348,35 @@
+    }
+    
+    RefPtr root = document->documentElement();
+    VisibleSelection fullDocumentSelection(VisibleSelection::selectionFromContentsOfNode(root.get()));
```
Is `root` guaranteed to be non-null here? Probably worth adding a null check.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2753,6 +2753,43 @@ - (void)_addAppHighlightInNewGroup:(BOOL)newGroup originatedInApp:(BOOL)originat
 
+- (NSUUID *)_enableTextIndicatorStylingAfterElementWithID:(NSString *)elementID
+{
+    RetainPtr nsuuid = [NSUUID UUID];
```
```suggestion
    RetainPtr nsUUID = [NSUUID UUID];
```

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -322,6 +348,35 @@
+    
+    auto liveRange = createLiveRange(*simpleRange);
+
+    m_textIndicatorStyleEnablementRanges.set(uuid, liveRange);
```
Nit - `add` is a tiny bit more efficient in the case where you don't expect to be replacing any existing entries. You can also use `WTFMove` here to prevent ref-churn (or simply remove the temporary variable)

```suggestion
    m_textIndicatorStyleEnablementRanges.add(uuid, createLiveRange(*simpleRange));
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2753,6 +2753,43 @@ - (void)_addAppHighlightInNewGroup:(BOOL)newGroup originatedInApp:(BOOL)originat
+#endif
+}
+
+- (void)_disableTextIndicatorStylingWithUUID:(NSUUID *)nsuuid
```
```suggestion
- (void)_disableTextIndicatorStylingWithUUID:(NSUUID *)nsUUID
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -396,6 +396,9 @@ for this property.
 - (void)_addAppHighlightInNewGroup:(BOOL)newGroup originatedInApp:(BOOL)originatedInApp WK_API_AVAILABLE(macos(12.0), ios(15.0));
 
+- (NSUUID *)_enableTextIndicatorStylingAfterElementWithID:(NSString *)elementID WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
+- (void)_disableTextIndicatorStylingWithUUID:(NSUUID *)nsuuid WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
```suggestion
- (void)_disableTextIndicatorStylingWithUUID:(NSUUID *)nsUUID WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2193,6 +2193,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
 #if ENABLE(UNIFIED_TEXT_REPLACEMENT)
     void removeTextIndicatorStyleForID(const WTF::UUID&);
+    void enableTextIndicatorStyleAfterElementWithID(const String& elementID, WTF::UUID);
```
```suggestion
    void enableTextIndicatorStyleAfterElementWithID(const String& elementID, const WTF::UUID&);
```

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -312,7 +312,33 @@
 void WebPage::getTextIndicatorForID(const WTF::UUID& uuid, CompletionHandler<void(std::optional<WebCore::TextIndicatorData>&&)>&& completionHandler)
 {
-    m_unifiedTextReplacementController->getTextIndicatorForID(uuid, WTFMove(completionHandler));
+    RefPtr range = m_unifiedTextReplacementController->rangeForUUID(uuid);
```
Just to double check — does this return `nullptr` in the event `uuid` is not tin the map? I wonder if it's just returning `Ref(HashTableEmptyValueType)`, and then implicitly converting that into a `RefPtr`

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -76,6 +76,7 @@
 #include <WebCore/PluginData.h>
 #include <WebCore/PointerCharacteristics.h>
 #include <WebCore/PointerID.h>
+#include <WebCore/Range.h>
```
Can we forward declare `Range`?

aprotyas commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -312,7 +312,33 @@
+
+    if (RefPtr localMainFrame = dynamicDowncast<LocalFrame>(m_page->mainFrame())) {
+        std::optional<TextIndicatorData> textIndicatorData;
+        constexpr OptionSet<TextIndicatorOption> textIndicatorOptions {
```
Nit: you can probably drop the template arg here.
```suggestion
        constexpr OptionSet textIndicatorOptions {
```

aprotyas reviewed (commented):
Looks good! Minus the style guide errors

megangardner committed:
Enable text indicator styling for specified ranges.
https://bugs.webkit.org/show_bug.cgi?id=272293
rdar://125359006

Reviewed by Wenson Hsieh.

Enable text indicator styling for specified ranges.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _enableTextIndicatorStylingAfterElementWithID:]):
(-[WKWebView _disableTextIndicatorStylingWithUUID:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::enableTextIndicatorStyleAfterElementWithID):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm:
(WebKit::UnifiedTextReplacementController::getTextIndicatorForID): Deleted.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getTextIndicatorForID):
(WebKit::WebPage::enableTextIndicatorStyleAfterElementWithID):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/277284@main
